guarddog/reporters/sarif.py

```diff
@@ -0,0 +1,185 @@
+import hashlib
+import json
+import yaml
+import os.path
+
+from guarddog.analyzer.metadata import get_metadata_detectors
+from guarddog.ecosystems import ECOSYSTEM
+
+
+def build_rules_help_list() -> dict:
+    rules_documentation = {}
+    for ecosystem in ECOSYSTEM:
+        rules = get_metadata_detectors(ecosystem)
+        for name, instance in rules.items():
+            detector_class = instance.__class__.__base__
+            rules_documentation[name] = detector_class.__doc__
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    semgrep_rules_base_dir = os.path.join(dir_path, "..", "analyzer", "sourcecode")
+    for file in os.listdir(semgrep_rules_base_dir):
+        if not file.endswith('.yml'):
+            continue
+        with open(os.path.join(semgrep_rules_base_dir, file), "r") as fd:
+            content = yaml.safe_load(fd)
+            for rule in content["rules"]:
+                text = rule["description"] if "description" in rule else rule["message"]
+                rules_documentation[rule["id"]] = text
+    return rules_documentation
+
+
+def get_sarif_log(runs):
+    """
+    https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#sariflog-object
+    """
+    return {
+        "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
+        "version": "2.1.0",
+        "runs": runs
+    }
+
+
+def get_run(results, driver):
+    """
+    https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#run-object
+    """
+    return {
+        "tool": {
+            "driver": driver
+        },
+        "results": results
+    }
+
+
+def get_driver(rules, ecosystem: str):
+    """
+    https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#toolcomponent-object
+    """
+    return {
+        "name": f"GuardDog-{ecosystem}",
+        "rules": rules
+    }
+
+
+def get_rule(rule_name: str, rules_documentation) -> dict:
+    """
+    https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#reportingdescriptor-object
+    """
+    message = rules_documentation[rule_name] if rules_documentation[rule_name] is not None else ""
+    return {
+        "id": rule_name,
+        "defaultConfiguration": {
+            "level": "warning"
+        },
+        "shortDescription": {
+            "text": f"GuardDog rule: {rule_name}"
```

Comment:
would it make sense (here or somewhere else in the guarddog output) to try and link to the exact rule implementation on GitHub?

Reply:
I think so, but this might be a stretch for now if were want to rethink the way rules are loaded in the next couple cycles